### PR TITLE
Tighten bitwise interval operations

### PR DESCRIPTION
**Pull request by timor**
Addresses #2170

- fixed: `interval-bitor` and `interval-bitxor` caused bit-growth
- improved: `interval-bitor` more exact about lower bounds

The added utility words could be used as a basis to make the other bitwise
interval operations more exact also.

### CODE REVIEW

File: basis/math/intervals/intervals-tests.factor [hunk 1]
```diff
@@ -384,3 +386,60 @@ commutative-ops [
...
+    16 <iota> dup [ bitor ] cartesian-map flatten
+    [ 0 15 [a,b] interval-contains? ] all?
+] unit-test
+
+: cartesian-bounds ( range range quot -- min max )
+    cartesian-map flatten minmax ; inline
+
+{ 0 15 } [ 16 <iota> dup [ bitor ] cartesian-bounds ] unit-test
+
+{ -8 7 } [ -8 7 1 <range> dup [ bitor ] cartesian-bounds ] unit-test
```
[the remark this answers is not in the document]
Comment (timor) [reply basis/math/intervals/intervals-tests.factor]: Correct.  They test the basic assumption behind the optimization.

File: basis/math/intervals/intervals.factor [hunk 2]
```diff
@@ -386,6 +386,9 @@ SYMBOL: incomparable
 : interval-nonnegative? ( i -- ? )
     from>> first 0 >= ;
 
+: interval-negative? ( interval -- ? )
```
[the remark this answers is not in the document]
Comment (timor) [reply basis/math/intervals/intervals.factor]: Yes, it means fully negative.  A number in such an interval is guaranteed to have the sign bit set.

File: basis/math/intervals/intervals-tests.factor [hunk 3]
```diff
@@ -384,3 +386,60 @@ commutative-ops [
...
+
+{ 6 15 } [ 5 15 1 <range> 6 15 1 <range> [ bitor ] cartesian-bounds ] unit-test
+
+{ -12 -1 } [ -16 -12 1 <range> -12 -2 1 <range> [ bitor ] cartesian-bounds ] unit-test
+
+{ -16 15 } [ -16 4 1 <range> -1 15 1 <range> [ bitor ] cartesian-bounds ] unit-test
+
+{ $[ 0 255 [a,b] ] } [ 0 255 [a,b] dup interval-bitor ] unit-test
```
[the remark this answers is not in the document]
Comment (timor) [reply basis/math/intervals/intervals-tests.factor]: Thanks for the hint!

File: basis/math/intervals/intervals.factor [hunk 4]
```diff
@@ -402,19 +405,71 @@ SYMBOL: incomparable
...
+    [ drop 1/0. ]
+    [ dup 0 > [ 1 + ] [ neg ] if next-power-of-2 ] if ;
+
+: bounds ( interval -- lower upper )
+    {
+        { full-interval [ -1/0. 1/0. ] }
+        [ interval>points [ first ] bi@ ]
+    } case ;
+
+: interval-negative? ( interval -- ? )
```
[the remark this answers is not in the document]
Comment (timor) [reply basis/math/intervals/intervals.factor]: Forgot to remove the private one, fixed.

**Comment by mrjbq7** [comment]
Hi @timor I wanted to follow up on these patches and see if you thought it was ready to merge?

**Comment by timor** [comment]
> Hi @timor I wanted to follow up on these patches and see if you thought it was ready to merge?

I think they are ready.  I have been using them for a bit now and don't think they were causing any trouble.

**Comment by timor** [comment]
Maybe documentation should be added before merging?

File: basis/math/intervals/intervals-tests.factor (hunk 1 above)
[the remark this answers is not in the document]
Comment (timor) [reply basis/math/intervals/intervals-tests.factor]: @mrjbq7 Should these be removed?

[the remark this answers is not in the document]
Comment (mrjbq7) [reply basis/math/intervals/intervals-tests.factor]: Sounds good, removed in 20c4e2feaaf338e4710019a2ebc1d8b108899abc.

**Comment by mrjbq7** [comment]
Merged, thanks!

**Pull request closed without merging**